File: src/map.ts
enum CellType {
  Empty,
  Player,
  Ghost,
  Wall,
  Pill
}

class Cell {
  type: CellType;

  constructor(type: CellType) {
    this.type = type;
  }
}

class Map {
  static parse(data: string): Map {
    let lines = data.split('\n');
    let map = new Map();
    let width = 0;
    lines.forEach(line => {
      line = line.trim();
      if (line.length > width) {
        width = line.length;
      }

      var row = [];
      map.cells.push(row);

      for (let i = 0; i < line.length; ++i) {
        var type = Map.parseCellType(line[i]);
        if (type != null) {
          row.push(new Cell(type));
        }
      }
    });

    map.cells.forEach(line => {
      while (line.length < width) {
        line.push(new Cell(CellType.Empty));
      }
    });

    return map;
  }

  static parseCellType(char: string) {
    switch (char) {
      case ' ':
        return CellType.Empty;
      case '#':
        return CellType.Wall;
      case 'o':
        return CellType.Player;
      case '.':
        return CellType.Pill;
      case 'g':
        return CellType.Ghost;
    }

    return null;
  }

  constructor() {
    this.cells = [];
  }

  cells: Cell[][];
}

export = Map;

File: src/main/kotlin/me/fornever/atoman/map/Map.kt
package me.fornever.atoman.map

private data class Map(val cells: List<List<Cell>>, val width: Int, val height: Int) {

    companion object {

        fun parse(data: String): Map {
            val lines = data.split('\n')
            val cells = mutableListOf<MutableList<Cell>>()
            var width = 0

            for (line in lines) {
                val trimmedLine = line.trim()
                if (trimmedLine == "") {
                    continue
                }

                if (trimmedLine.length > width) {
                    width = trimmedLine.length
                }

                val row = mutableListOf<Cell>()
                cells.add(row)

                for (element in trimmedLine) {
                    val type = parseCellType(element)
                    if (type != null) {
                        row.add(Cell(type))
                    }
                }
            }

            for (row in cells) {
                while (row.size < width) {
                    row.add(Cell(CellType.Empty))
                }
            }

            return Map(
                cells,
                width,
                cells.size
            )
        }

        private fun parseCellType(char: Char): CellType? =
            when (char) {
                ' ' -> CellType.Empty
                '#' -> CellType.Wall
                'o' -> CellType.Player
                '.' -> CellType.Pill
                'g' -> CellType.Ghost
                else -> null
            }
    }
}
